application/src/main/java/org/togetherjava/tjbot/commands/free/FreeHelpChannelsRoutine.java

```diff
@@ -0,0 +1,54 @@
+package org.togetherjava.tjbot.commands.free;
+
+import net.dv8tion.jda.api.JDA;
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.Routine;
+import java.time.OffsetDateTime;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Routine that automatically marks inactive help channels as free.
+ * <p>
+ * The routine runs every 10 minutes, and checks for help channels that have been inactive for more
+ * than 2 hours.
+ * </p>
+ */
+public class FreeHelpChannelsRoutine implements Routine {
+    private final ChannelMonitor channelMonitor;
+
+
+    /**
+     * Creates an instance of {@link FreeHelpChannelsRoutine}.
+     *
+     * @param channelMonitor the {@link ChannelMonitor} to use for marking channels as free
+     */
+    public FreeHelpChannelsRoutine(@NotNull ChannelMonitor channelMonitor) {
+        this.channelMonitor = channelMonitor;
+    }
+
+    @Override
+    public void runRoutine(@NotNull JDA jda) {
+        jda.getTextChannels()
+            .stream()
+            .filter(channel -> channelMonitor.isMonitoringChannel(channel.getIdLong())
+                    && channelMonitor.isChannelBusy(channel.getIdLong()))
+            .forEach(channel -> channel.retrieveMessageById(channel.getLatestMessageId())
+                .queue(latestMessage -> {
```

Comment:
and I don't know, but if possible look for a way to refactor this?
This is also very hard to read

Reply:
can you be a bit more specific? do you suggest to get the channels first, put them in a some kind of list, then iterate over them, or something very different?

Reply:
2 points you could take a look at

- getLatestMessageId return's the last message **send** their ID, so if a message gets removed, it's still that ID, so this will error
- Take a look at TimeUtil#getTimeCreated, an ID contains the creation date already

Reply:
24331a6c1f9f6796f890c40ba5e820cacb2c3b44